src/tss2-tcti/tcti-swtpm.c

```diff
@@ -0,0 +1,640 @@
+ *  > execute in another thread that polls on a cancel
+ *  > flag
+ */
+TSS2_RC
+tcti_swtpm_cancel (
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    (void) (tctiContext);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+TSS2_RC
+tcti_swtpm_set_locality (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    uint8_t locality)
+{
+    TSS2_TCTI_SWTPM_CONTEXT *tcti_swtpm = tcti_swtpm_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_swtpm_down_cast (tcti_swtpm);
+    TSS2_RC rc;
+
+    if (tcti_swtpm == NULL) {
+        return TSS2_TCTI_RC_BAD_CONTEXT;
+    }
+    rc = tcti_common_set_locality_checks (tcti_common);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    rc = tcti_control_command (tctiContext, CMD_SET_LOCALITY,
+                               &locality, sizeof(locality),
+                               NULL, NULL, 0);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOG_ERROR("Failed to set locality: 0x%" PRIx32, rc);
+        return rc;
+    }
+
+    tcti_common->locality = locality;
+    return TSS2_RC_SUCCESS;
+}
+
+/*
+ * For compatibility to the mssim platform commands issued by some tests:
+ * Mapping mssim platform commands to swtpm control channel commands //TODO
+ */
+TSS2_RC tcti_platform_command (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    UINT32 cmd)
+{
+    LOG_TRACE("Mapping mssim platform command: %" PRIu32, cmd);
+
+    if (cmd == 1) {         //MS_SIM_POWER_ON
+        return tcti_swtpm_reset(tctiContext);
+    } else if (cmd == 2) {  // MS_SIM_POWER_OFF
+        return TSS2_RC_SUCCESS;
+    } else if (cmd == 9) {  // MS_SIM_CANCEL_ON
+        // to SKIP sapi-command-cancel.int.c
+        return TSS2_TCTI_RC_BAD_CONTEXT;
+    } else if (cmd == 10) {  // MS_SIM_CANCEL_OFF
+        return TSS2_RC_SUCCESS;
+    } else {
+        return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+    }
+}
+
+TSS2_RC
+tcti_swtpm_get_poll_handles (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    TSS2_TCTI_POLL_HANDLE *handles,
+    size_t *num_handles)
+{
+    (void)(tctiContext);
+    (void)(handles);
+    (void)(num_handles);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+void
+tcti_swtpm_finalize(
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    TSS2_TCTI_SWTPM_CONTEXT *tcti_swtpm = tcti_swtpm_context_cast (tctiContext);
+
+    if (tcti_swtpm == NULL) {
+        return;
+    }
+
+    socket_close (&tcti_swtpm->tpm_sock);
+}
+
+TSS2_RC
+tcti_swtpm_receive (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    size_t *response_size,
+    unsigned char *response_buffer,
+    int32_t timeout)
+{
+    TSS2_TCTI_SWTPM_CONTEXT *tcti_swtpm = tcti_swtpm_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_swtpm_down_cast (tcti_swtpm);
+    TSS2_RC rc;
+    int ret;
+
+    if (tcti_swtpm == NULL) {
+        return TSS2_TCTI_RC_BAD_CONTEXT;
+    }
+    rc = tcti_common_receive_checks (tcti_common, response_size);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    if (timeout != TSS2_TCTI_TIMEOUT_BLOCK) {
```

Comment:
Could we get async I/O working ?
That's gonna be needed by FAPI and potentially others in the future.

Reply:
AFAIK, `libtpms` does not support async rx. It does not even support separated tx/rx but only one call 
```c
TPM_RESULT TPMLIB_Process(unsigned char **respbuffer, uint32_t *resp_size,
                          uint32_t *respbufsize,
                          unsigned char *command, uint32_t command_size);
```